test: Relay test server

timfish opened the pull request:
This PR relies on some [minor changes](https://github.com/getsentry/relay/pull/1151) to Relay that add an endpoint which can be used when it's run in Capture mode. This mode was [added to facilitate unit testing](https://github.com/getsentry/relay/pull/238). The additional endpoint allows polling for all captured events.

This PR adds the `@sentry-internal/test-server` module with the `RelayTestServer` class which runs the Relay executable and polls for and collects new events. If you build my linked Relay PR and set `RELAY_LOCAL_PATH` to point at the built executable, you can see it in action!

I've run the Electron SDK e2e tests against this test server and the only issue seems to be that sessions are missing. The existing Relay capture mode doesn't hook into sessions yet.

AbhiPrasad commented on packages/test-server/src/types.ts:
```diff
@@ -0,0 +1,189 @@
+
+export type ServerEvent = OkEvent | ErrEvent;
+
+export interface Event {
```
Tbh these types should just be in `@sentry/types` but we can deal with that afterwards.

timfish replied:
I auto generated most of these types from all the data submitted through the test server using https://quicktype.io and made some minor fixes.

They differ slightly from `@sentry/types` in places because Relay does some normalisation. The only example that springs to mind is `tags` becoming `string[][]` after Relay rather than `Record<string, string>`.

AbhiPrasad replied:
Ah yes you are right about the small typing mismatches. I explored the JSON schema relay exported a while back https://github.com/getsentry/sentry-javascript/pull/2802 and hit those.

AbhiPrasad replied:
We can codemod those small differences out 😄

AbhiPrasad commented:
@timfish I’m adding this as part of our release stability milestone for the js sdk, we’ll have a discussion about this tommorow and then let’s chat about next steps.

timfish commented:
A couple of points worth discussing:
- What envelopes need supporting to cover SDK tests? I'm guessing we need to add session and session aggregate
- Is the order of event/session envelopes important for any tests? It probably shouldn't be since I don't think Sentry cares about the order received. Capture mode doesn't currently maintain order.
- Who/where currently uses the Relay capture mode? I don't want to break existing usage but couldn't find it being used anywhere.

timfish commented:
WIP

timfish commented:
I've updated the Relay PR to support all envelope types rather than just Events and updated this PR to reflect the changes. Still need to test that it's actually working correctly and add some more tests.

AbhiPrasad commented:
Let's close this for now, we can always try this again in the future.
